### ISSUE TITLE
Positioning inside container-type: inline-size is off when using position: fixed on the popup

### ISSUE THREAD

**Issue by IvandeWolf** [floating-ui/floating-ui#3067]
**Describe the bug**

When using the `computePosition()` function in the following context the popup is placed off
(see also https://codesandbox.io/p/sandbox/funny-glade-jfr7kl):

- The button and popup element are located inside a div somewhere in the DOM tree with the CSS property: `container-type: inline-size`
- Using strategy `fixed`
- The popup element has CSS property `position: fixed`. This is because we want the popup to be visible and overflowing even in a small side container for example.

**To Reproduce**

See code sandbox: https://codesandbox.io/p/sandbox/funny-glade-jfr7kl

**Expected behavior**

The popup should be placed below the button.

**Screenshots**

![image](https://github.com/user-attachments/assets/4be152b0-3f55-41ae-bb6f-08791a15baac)

**Additional context**

I have debugged the code and found the following code to be executed which might be 'a problem' but I also might be doing something weird. (Found in dom/src/utils/getRectRelativeToOffsetParent.ts)

https://github.com/floating-ui/floating-ui/blob/master/packages/dom/src/utils/getRectRelativeToOffsetParent.ts#L23
```ts
/**
 * Here we get the coordinator for the element and this actually already returns
 * the correct x and y coordinates for the popup placement.
 */
const rect = getBoundingClientRect(element, true, isFixed, offsetParent);
```

https://github.com/floating-ui/floating-ui/blob/master/packages/dom/src/utils/getRectRelativeToOffsetParent.ts#L36
```ts
/**
 * We enter this if-statement, because the offsetParent is the div
 * with the container-type: inline-size and that is indeed an element.
 */
if (isOffsetParentAnElement) {
    /**
     * This returns the rect values of the offsetParent.
     * Which means the x and y coordinator of the div with container-type: inline-size.
     */
    const offsetRect = getBoundingClientRect(
      offsetParent,
      true,
      isFixed,
      offsetParent,
    );
    offsets.x = offsetRect.x + offsetParent.clientLeft;
    offsets.y = offsetRect.y + offsetParent.clientTop;
  } else if (documentElement) {
```

And these offsets are being used here:

https://github.com/floating-ui/floating-ui/blob/master/packages/dom/src/utils/getRectRelativeToOffsetParent.ts#L65
```ts
/**
 * Here we substract the offsets from the x and y coordinates which make the position move back.
 * While the rect.left and rect.top where already the correct placement for the popup.
 */
const x = rect.left + scroll.scrollLeft - offsets.x - htmlX;
const y = rect.top + scroll.scrollTop - offsets.y - htmlY;
```

So based on this analysis I feel like we substract the offsets but why? I feel like there is a good scenario where the offset should be used, but this might be one where it shouldn't. Or is there something else in the setup that is wrong?

**Comment by kwarr128** [comment]
I'm having the same issue, but only in the new version of Chrome. I had some user reports about it for my application, but I wasn't having any issues until i updated to Chrome 129. It worked fine on Chrome 128, as well as Firefox 131 and Safari 16.3 (i need to update). I'm on Mac

**Comment by mikehallgenesys** [comment]
Same issue using `size` as well.

**Comment by atomiks** [comment]
The `container-type` property creates a "containing block", at least in Chrome 128 and below. Either Chrome 129 has a regression for this, or the spec changed and it was updated. I'm not sure where to look to see if that's the case. [On MDN, `container-type` was added last year](https://github.com/mdn/content/commit/ff4601c1b62ccdaadb4c15040f0d2429b3764124) but remains to this day.

Either way, it's going to be a pain to handle browser versions to get it to work reliably. Issues with containing blocks causing positioning issues comes up frequently because browsers randomly change how a block is created (or regress), and different browsers handle things differently (i.e. if they start matching a spec or Chrome, positioning breaks just from updating the browser version and there's nothing we can do)

**Comment by IvandeWolf** [comment]
Good catch! I didn't think to check different browsers and their versions, but viewing the same page in firefox indeed works correctly.

**Comment by IvandeWolf** [comment]
Alright, after reading the blog it comes down to this:

Prior to chrome 129, `container-type: inline-size` creates an implicit containing block for absolutely positioned children. So if you tell a child to place it `left: 0` it will be against the border of the container.

Apparently according to the Chromium team this is a design mistake so they 'fixed' it. But all other major browsers still follow the pre-Chrome 129 behavior.

### Workaround
What works for me is the to add `contain: layout;` to the container div to explicitly add a containing block.
I also updated the CodeSandbox to show that it works: https://codesandbox.io/p/sandbox/funny-glade-jfr7kl

**Comment by atomiks** [comment]
I'm kind of leaning toward leaving it stay broken, and even removing the check for Firefox and Safari too so it's consistently broken across all browsers. This might sound weird, but there are two reasons: 

1. Determining that >=129 version of Chromium is being used (including all browsers based on it, like Edge) might be hard. If Chromium forces the same versioning with the [`userAgentData API` or `navigator.userAgent`](https://developer.mozilla.org/en-US/docs/Web/API/Navigator/userAgentData) then this point is not really valid though.

2. But consider Firefox and Safari. One day randomly, like what happened here, when the spec is changed and they update, positioning will suddenly break again for those two engines. But if you force `contain: layout` to be used when adding `container-type: inline-size`, then you'll never have to encounter that. By leaving it consistently broken, future issues get solved despite the initial cost of needing to workaround it being broken unexpectedly initially. Once Firefox and Safari have both updated, and we know the versions they did it in, then we can "fix" it by default.

**Comment by joaopslins** [comment]
> I'm kind of leaning toward leaving it stay broken, and even removing the check for Firefox and Safari too so it's consistently broken across all browsers. 

>  Once Firefox and Safari have both updated, and we know the versions they did it in, then we can "fix" it by default.

Agree with this take.

Until that happens though, I wonder if we could have some note in the documentation for the time being. I just spent a couple of hours trying to debug this behavior and it wasn't obvious about what's happening. 😅 

I think by having a warning in the docs could help future devs in the future.

**Comment by symmetriq** [comment]
`contain: layout` on the parent of the floating element (a dropdown menu, in my case) works for me, but in some cases that causes the menu to be clipped by the container, so I also needed to add `z-index` to the container with the same value I use for the menu. To avoid creating other z-index issues, I apply the z-index to the container only when the menu is actually open.

A better overall strategy for floating elements is to just render them as direct children of `<body>`, and use the appropriate `aria` attributes to establish their relationship to the element they're anchored to.

**Comment by zhijunsong-cohere** [comment]
Hey, would we consider consider reconstruct the element popover trigger container like this:
so that both button and popover modal would be wrapped under the same div wrapper, and change the position:absolute; transform: none; top: buttonheight+gap or something;

```
position:absolute; 
transform: none;
top: buttonheight + gap size;
```

<img width="1149" height="606" alt="Image" src="https://github.com/user-attachments/assets/b925605e-6831-4714-911b-12f6e4473cad" />

**Comment by cvx** [comment]
> Once Firefox and Safari have both updated, and we know the versions they did it in, then we can "fix" it by default.

It seems that all browsers are on the same page now: https://wpt.fyi/results/?label=master&label=stable&product=chrome&product=firefox&product=safari&aligned&q=no-layout-containment

I believe that the change for Safari was: https://github.com/WebKit/WebKit/commit/e0a77700ca7ba9b67cffb56db95f76a22d2bb631 (landed in [Safari Technology Preview 206](https://webkit.org/blog/16186/release-notes-for-safari-technology-preview-206/), and then in [Safari 18.4](https://webkit.org/blog/16574/webkit-features-in-safari-18-4/) in March 2025)

And for Firefox it was: https://hg-edge.mozilla.org/mozilla-central/rev/9285e195af8e (landed in Firefox 133 on November 2024?)
